jans-cedarling/cedarling/src/authz/request.rs

```diff
@@ -14,15 +14,28 @@ pub struct Request {
     /// cedar_policy action
     pub action: String,
     /// cedar_policy resource data
-    pub resource: ResourceData,
+    pub resource: EntityData,
     /// context to be used in cedar_policy
     pub context: serde_json::Value,
 }
 
-/// Cedar policy resource data
+/// Box to store authorization data, with any additional principals
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub struct RequestUnverified {
+    /// Contains the JWTs that will be used for the AuthZ request
+    pub principals: Vec<EntityData>,
```

Comment:
Have you considered that approach where instead of having two structs, we would just have one `Request` struct then have an `enum` for the principals instead?

```rs
pub enum RequestPrincipal {
    Tokens(HashMap<String, String>),
    Cedar(Vec<EntityData>),
}
```

```suggestion
    /// Contains the principals that will be used for the AuthZ request
    pub principals: RequestPrincipal,
```

In the target issue, you mentioned that you opted to have two functions instead of one because of the [Interface Segregation Principal (ISP)](https://reflectoring.io/interface-segregation-principle/). From what I read, the essence of ISP is:

>  a client should not be exposed to methods it doesn’t need.

If we have two functions, and a client only want to use the JWT ***or*** the Cedar inputs but ***not both***, aren't we essentially exposing them two at least one method they don't need?

Reply:
I think having a single function would also simplify the error hierarchy.

```rs
pub enum AuthorizeError {
    // These two functions wraps `cedar_policy::RequestValidationError` but for `Workload` and `User` respectively
    #[error("The request for `Workload` does not conform to the schema: {0}")]
    WorkloadRequestValidation(cedar_policy::RequestValidationError),
    #[error("The request for `User` does not conform to the schema: {0}")]
    UserRequestValidation(cedar_policy::RequestValidationError),

    // This wraps `cedar_policy::RequestValidationError` but for any entity type respectively
    #[error(transparent)]
    UnverifiedPrincipalRequestValidation(#[from] UnverifiedPrincipalRequestValidationError),
}

#[derive(Debug, derive_more::Error, derive_more::Display)]
#[display("The request for `{principal}` does not conform to the schema: {err}")]
pub struct UnverifiedPrincipalRequestValidationError {
    /// Principal name
    principal: String,
    /// Error value
    err: cedar_policy::RequestValidationError,
}
```

If we just have one interface, we can have a single error for building principals.

```rs
pub enum AuthorizeError {
    #[error(transparent)]
    InvalidPrincipal(InvalidPrincipal)
}

#[derive(Debug, derive_more::Error, derive_more::Display)]
#[display("the request for `{principal}` does not conform to the schema: {err}")]
pub struct InvalidPrincipal {
    /// Principal name
    principal: String,
    /// Error value
    err: cedar_policy::RequestValidationError,
}
```

Reply:
the naming was discussed [here](https://github.com/JanssenProject/jans/issues/11000#issuecomment-2705101286)

Your Idea with enum
```
pub enum RequestPrincipal {
    Tokens(HashMap<String, String>),
    Cedar(Vec<EntityData>),
}
```
is OK, but it can be almost impossible to implement on other languages.


___
about error hierarchy, yeah, we can rename `UnverifiedPrincipalRequestValidationError` to `InvalidPrincipal` and remove  `WorkloadRequestValidation` and  `UserRequestValidation`

Reply:
> the naming was discussed here. Maybe you prefer suffix unsigned?

yeah, `authorize_unsigned` is better.

> is OK, but it can be almost impossible to implement on other languages.

hmm... that's true. i'm thinking the other languages should just send a map then our bindings will process it but yeah... having two functions would probably the simpler way to go about it.

Reply:
method renamed 6d6e885fb6f1b2d73d605e143cddb168fc486aec

Reply:
Updated handling error 66127932bf33c5f657acb59185a96b69d017cbfe